src/librustc_mir/transform/check_consts/validation.rs

```diff
@@ -127,37 +126,47 @@ impl Deref for Validator<'_, 'mir, 'tcx> {
     }
 }
 
+pub fn compute_indirectly_mutable_locals<'mir, 'tcx>(
+    item: &Item<'mir, 'tcx>,
+) -> RefCell<IndirectlyMutableResults<'mir, 'tcx>> {
+    let dead_unwinds = BitSet::new_empty(item.body.basic_blocks().len());
+
+    let indirectly_mutable_locals = old_dataflow::do_dataflow(
+        item.tcx,
+        item.body,
+        item.def_id,
+        &[],
+        &dead_unwinds,
+        old_dataflow::IndirectlyMutableLocals::new(item.tcx, item.body, item.param_env),
+        |_, local| old_dataflow::DebugFormatted::new(&local),
+    );
+
+    let indirectly_mutable_locals = old_dataflow::DataflowResultsCursor::new(
+        indirectly_mutable_locals,
+        item.body,
+    );
+
+    RefCell::new(indirectly_mutable_locals)
+}
+
 impl Validator<'a, 'mir, 'tcx> {
-    pub fn new(item: &'a Item<'mir, 'tcx>) -> Self {
+    pub fn new(
+        item: &'a Item<'mir, 'tcx>,
+        indirectly_mutable_locals: &'a RefCell<IndirectlyMutableResults<'mir, 'tcx>>,
+    ) -> Self {
         let dead_unwinds = BitSet::new_empty(item.body.basic_blocks().len());
 
-        let indirectly_mutable_locals = old_dataflow::do_dataflow(
-            item.tcx,
-            item.body,
-            item.def_id,
-            &[],
-            &dead_unwinds,
-            old_dataflow::IndirectlyMutableLocals::new(item.tcx, item.body, item.param_env),
-            |_, local| old_dataflow::DebugFormatted::new(&local),
-        );
-
-        let indirectly_mutable_locals = old_dataflow::DataflowResultsCursor::new(
-            indirectly_mutable_locals,
-            item.body,
-        );
-        let indirectly_mutable_locals = Rc::new(RefCell::new(indirectly_mutable_locals));
-
         let needs_drop = FlowSensitiveResolver::new(
             NeedsDrop,
             item,
-            indirectly_mutable_locals.clone(),
+            indirectly_mutable_locals,
             &dead_unwinds,
         );
 
         let has_mut_interior = FlowSensitiveResolver::new(
             HasMutInterior,
             item,
-            indirectly_mutable_locals.clone(),
+            indirectly_mutable_locals,
             &dead_unwinds,
         );
```

Comment:
I wonder if it would be possible to couple these two `FlowSensitiveResolver`s so you wouldn't need two references to `indirectly_mutable_locals`.
Or maybe have a method to build each on the fly (in that case they'd only have references to data stored elsewhere).

Reply:
We could store `indirectly_mutable_locals` in `Qualifs` if we do something like this:

```rust
struct Qualifs<...> {
    indirectly_mutable_locals: DataflowResultsCursor::<...>,
    needs_drop: FlowSensitiveResolver<..., NeedsDrop>,
    has_mut_interior: FlowSensitiveResolver<..., HasMutInterior>,
}

impl QualifResolver<NeedsDrop> for Qualifs<...> {
     /* ... */
}

impl QualifResolver<HasMutInterior> for Qualifs<...> {
     /* ... */
}
```

Then we do intersection with `indirectly_mutable_locals` and masking with `in_any_value_of_ty` in the latter impls. Is this preferable to the existing code? It does get rid of a `RefCell`, and the only down side is that you need to use turbofish to get qualifs a la `QualifResolver::<NeedsDrop>::contains(&mut self.qualifs, local)`. I can probably add a helper method to fix this.

Reply:
The current impl seems fine. We can fine tune in separate (smaller) PRs